src/libostree/ostree-sysroot-deploy.c

```diff
@@ -2404,50 +2415,301 @@ ostree_sysroot_deploy_tree (OstreeSysroot     *self,
+  g_variant_builder_add (builder, "{sv}", "target",
+                         serialize_deployment_to_variant (deployment));
+
+  if (merge_deployment)
+    g_variant_builder_add (builder, "{sv}", "merge-deployment",
+                           serialize_deployment_to_variant (merge_deployment));
+
+  if (override_kernel_argv)
+    g_variant_builder_add (builder, "{sv}", "kargs",
+                           g_variant_new_strv ((const char *const*)override_kernel_argv, -1));
+
+  char *dnbuf = strdupa (_OSTREE_SYSROOT_RUNSTATE_STAGED);
+  const char *parent = dirname (dnbuf);
+  if (!glnx_shutil_mkdir_p_at (AT_FDCWD, parent, 0755, cancellable, error))
+    return FALSE;
+
+  g_autoptr(GVariant) state = g_variant_ref_sink (g_variant_builder_end (builder));
+  if (!glnx_file_replace_contents_at (AT_FDCWD, _OSTREE_SYSROOT_RUNSTATE_STAGED,
+                                      g_variant_get_data (state), g_variant_get_size (state),
+                                      GLNX_FILE_REPLACE_NODATASYNC,
+                                      cancellable, error))
+    return FALSE;
+
+  if (!_ostree_sysroot_reload_staged (self, error))
+    return FALSE;
+
+  return TRUE;
+}
+
+/* Invoked at shutdown time by ostree-deploy-staged.service */
+gboolean
+_ostree_sysroot_deploy_staged (OstreeSysroot *self,
+                               GCancellable  *cancellable,
+                               GError       **error)
+{
+  /* The service has a ConditionPathExists=/run/ostree/staged-deployment */
+  if (!self->staged_deployment)
+    return glnx_throw (error, "No staged deployment");
+  g_assert (self->staged_deployment_data);
+
+  g_autoptr(OstreeDeployment) merge_deployment = NULL;
+  g_autoptr(GVariant) merge_deployment_v = NULL;
+  if (g_variant_lookup (self->staged_deployment_data, "merge-deployment", "@a{sv}",
+                        &merge_deployment_v))
+    {
+      merge_deployment =
+        _ostree_sysroot_deserialize_deployment_from_variant (merge_deployment_v, error);
+      if (!merge_deployment)
+        return FALSE;
+    }
+  g_autofree char **kargs = NULL;
+  g_variant_lookup (self->staged_deployment_data, "kargs", "^a&s", &kargs);
+
+  /* Unlink the staged state now; if we're interrupted in the middle,
+   * we don't want e.g. deal with the partially written /etc merge.
    */
-  if (!write_origin_file_internal (self, sepolicy, new_deployment, NULL,
-                                   GLNX_FILE_REPLACE_NODATASYNC,
-                                   cancellable, error))
+  if (!glnx_unlinkat (AT_FDCWD, _OSTREE_SYSROOT_RUNSTATE_STAGED, 0, error))
+    return FALSE;
+
+  if (!sysroot_finalize_deployment (self, self->staged_deployment, NULL, merge_deployment,
+                                    cancellable, error))
+    return FALSE;
+
+  /* TODO: Proxy across flags too? */
+  OstreeSysrootSimpleWriteDeploymentFlags flags = 0;
+  if (!ostree_sysroot_simple_write_deployment (self, ostree_deployment_get_osname (self->staged_deployment),
```

Comment:
Question/minor: noticed in rpm-ostree, seems like we had `rpmostree_syscore_write_deployment` to add some flags for livefs and rollback, do we need to consider a bit here to make it easier to integrate this into rpm-ostree in the future?(if needed :p)

Reply:
I think you're right that integrating this nicely will require pulling down some of that logic.